src/nias_cpp/interfaces/inner_products.h

```diff
@@ -0,0 +1,155 @@
+#ifndef NIAS_CPP_INTERFACES_INNER_PRODUCTS_H
+#define NIAS_CPP_INTERFACES_INNER_PRODUCTS_H
+
+#include <memory>
+#include <optional>
+
+#include <nias_cpp/concepts.h>
+#include <nias_cpp/indices.h>
+#include <nias_cpp/interfaces/vectorarray.h>
+#include <nias_cpp/type_traits.h>
+#include <pybind11/numpy.h>
+
+namespace nias
+{
+
+
+template <floating_point_or_complex F>
+class SesquilinearFormInterface
+{
+   public:
+    using ScalarType = F;
+
+    SesquilinearFormInterface() = default;
+    SesquilinearFormInterface(const SesquilinearFormInterface&) = default;
+    SesquilinearFormInterface(SesquilinearFormInterface&&) = default;
+    SesquilinearFormInterface& operator=(const SesquilinearFormInterface&) = default;
+    SesquilinearFormInterface& operator=(SesquilinearFormInterface&&) = default;
+    virtual ~SesquilinearFormInterface() = default;
+
+    /**
+     * \brief Apply the sesquilinear form
+     *
+     * If \c pairwise is \c true, the form is applied to each pair of vectors in the arrays and
+     * the result has the same length as the input arrays (which have to have the same size in this case).
+     * If \c pairwise is \c false, the form is applied to each vector in the first array with each vector
+     * of the second array, and the result has size <tt>left.size() * right.size()</tt>
+     * (where the <tt>i * right.size() + j</tt>-th element contains the result of applying the form to
+     * the <tt>i</tt>-th vector of the left array with the <tt>j</tt>-th vector of the right array).
+     */
+    virtual std::vector<ScalarType> apply(
+        const VectorArrayInterface<ScalarType>& left, const VectorArrayInterface<ScalarType>& right,
+        bool pairwise = false, const std::optional<Indices>& left_indices = std::nullopt,
+        const std::optional<Indices>& right_indices = std::nullopt) const = 0;
```

Comment:
Following Tuesdays discussion I think we wanted to have something like
```suggestion
    virtual std::vector<std::vector<ScalarType>> apply(
        const VectorArrayInterface<ScalarType>& left, const VectorArrayInterface<ScalarType>& right,
        const std::optional<Indices>& left_indices = std::nullopt,
        const std::optional<Indices>& right_indices = std::nullopt) const = 0;

    virtual std::vector<ScalarType> apply_pairwise(
        const VectorArrayInterface<ScalarType>& left, const VectorArrayInterface<ScalarType>& right,
        const std::optional<Indices>& left_indices = std::nullopt,
        const std::optional<Indices>& right_indices = std::nullopt) const = 0;
```